File: Data Structures & Algorithms/maximum-subarray/submission-1.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maxnum = nums[0];
        if(nums.size() == 1) return nums[0];
        int curr = 0;
        for(int i = 0;i < nums.size(); i++){
            curr += nums[i];
            maxnum = max(maxnum, curr);
            if(curr < 0) {
                curr = 0;
            }
        }
        return maxnum;
    }
};

File: Data Structures & Algorithms/maximum-product-subarray/submission-1.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        cur_max= cur_min= res = nums[0]

        for num in nums[1:]:
            if num < 0:
                cur_max, cur_min = cur_min, cur_max
            cur_max = max(num, cur_max * num)
            cur_min = min(num, cur_min * num)

            res = max(res, cur_max)
        return res
        # cursum, curmin, maxsum = nums[0]
        # # maxsum = nums[0]
        # for num in nums[1:]:
        #     if cursum * num < cursum:
        #         cursum = 1
        #     cursum = max(cursum * num, cursum)
        #     maxsum = max(cursum, maxsum)
        # return maxsum